### PR TITLE
Add support for reading and writing <volume> tags

### LINKED ISSUE #34: Allow to write volume tag
When supplementing an NML file with volume data, webKnossos includes a volume tag in the NML file, like `<volume id="1" location="data.zip" fallbackLayer="segmentation" />`.

wknml should allow to write this volume tag as well.

### PR DESCRIPTION
**Pull request by hotzenklotz**
## Changes
- PR adds supports for reading and writing volume tags from NML files:
```
<volume id="1" location="data.zip" fallbackLayer="segmentation" />
```
- extended type annotations
- added more doc strings for Documentation

## Issues
Fixes #34

### CODE REVIEW

File: docs/wknml.md [hunk 1]
```diff
@@ -197,3 +229,119 @@ Writes an NML object to a file on disk.
...
+
+```python
+globalize_tree_ids(group_dict: Dict[str, List[nx.Graph]])
+```
+
+A utility to in-place re-assign new and globally unqiue IDs to all Tree objects. Starts with ID 1
+
+**Arguments**:
+
+- `group_dict` _Dict[str, List[nx.Graph]]_ - A mapping of group names to a list of tree as NetworkX graph objects
```
Comment (daniel-wer) [inline docs/wknml.md:264]:
```suggestion
- `group_dict` _Dict[str, List[nx.Graph]]_ - A mapping of group names to a list of trees as NetworkX graph objects
```

File: docs/wknml.md [hunk 2]
```diff
@@ -197,3 +229,119 @@ Writes an NML object to a file on disk.
...
+<a name="wknml.nml_generation.globalize_node_ids"></a>
+#### globalize\_node\_ids
+
+```python
+globalize_node_ids(group_dict: Dict[str, List[nx.Graph]])
+```
+
+A utility to in-place re-assign new and globally unqiue IDs to all Node objects. Edges are updated accordingly. Starts with ID 1.
+
+Note: Does not update any `Comment`s or `BranchPoint`s referencing these nodes.
```
Comment (daniel-wer) [inline docs/wknml.md:275]: In this state (where the trees are nx.Graphs), branchpoints and comments are indicated by a `branchpoint` or `comment` attribute at the respective node, so there is no need to update any IDs. The separate branchpoints and comments lists are only generated from the nx.Graphs, later.

File: docs/wknml.md [hunk 3]
```diff
@@ -197,3 +229,119 @@ Writes an NML object to a file on disk.
...
+
+**Arguments**:
+
+- `group_dict` _Dict[str, List[nx.Graph]]_ - A mapping of group names to a list of tree as NetworkX graph objects
```
Comment (daniel-wer) [inline docs/wknml.md:279]:
```suggestion
- `group_dict` _Dict[str, List[nx.Graph]]_ - A mapping of group names to a list of trees as NetworkX graph objects
```

File: docs/wknml.md [hunk 4]
```diff
@@ -197,3 +229,119 @@ Writes an NML object to a file on disk.
...
+
+```python
+generate_nml(tree_dict: Union[List[nx.Graph], Dict[str, List[nx.Graph]]], parameters: Dict[str, Any] = {}, globalize_ids: bool = True, volume: Optional[Dict[str, Any]] = None) -> NML
+```
+
+A utility to convert a [NetworkX graph object](https://networkx.org/) into wK NML skeleton annotation object. Accepts both a simple list of multiple skeletons/trees or a dictionary grouping skeleton inputs.
+
+**Arguments**:
+
+- `tree_dict` _Union[List[nx.Graph], Dict[str, List[nx.Graph]]]_ - A list of wK tree-like structures as NetworkX graphs or a dictionary of group names and same list of NetworkX tree objects.
```
Comment (daniel-wer) [inline docs/wknml.md:292]:
```suggestion
- `tree_dict` _Union[List[nx.Graph], Dict[str, List[nx.Graph]]]_ - A list of wK tree-like structures as NetworkX graphs or a dictionary of group names and lists of NetworkX tree objects.
```

File: docs/wknml.md [hunk 5]
```diff
@@ -197,3 +229,119 @@ Writes an NML object to a file on disk.
...
+- `parameters` _Dict[str, Any]_ - A dictionary representation of the skeleton annotation metadata. See `NMLParameters` for accepted attributes.
+- `globalize_ids` _bool = True_ - An option to re-assign new, globally unique IDs to all skeletons. Default: `True`
+- `volume` _Optional[Dict[str, Any]] = None_ - A dictionary representation of a reference to wK a volume annotation. See `Volume` object for attributes.
```
Comment (daniel-wer) [inline docs/wknml.md:295]:
```suggestion
- `volume` _Optional[Dict[str, Any]] = None_ - A dictionary representation of a reference to a wK volume annotation. See `Volume` object for attributes.
```

File: docs/wknml.md [hunk 6]
```diff
@@ -197,3 +229,119 @@ Writes an NML object to a file on disk.
...
+  
+
+**Returns**:
+
+- `nml` _NML_ - A wK NML skeleton annotation object
+  1. A dictionary with group names as keys and list of all respective NML trees as values
+  2. A dictionary representation of the NML metadata parameters
```
Comment (daniel-wer) [inline docs/wknml.md:302]:
```suggestion
```

Looks like this is a leftover as these are not returned by `generate_nml`.

File: docs/wknml.md [hunk 7]
```diff
@@ -197,3 +229,119 @@ Writes an NML object to a file on disk.
...
+
+<a name="wknml.nml_generation.generate_graph"></a>
+#### generate\_graph
+
+```python
+generate_graph(nml: NML) -> Tuple[Dict[str, List[nx.Graph]], Dict[Text, any]]
+```
+
+A utility to convert wK NML object into a [NetworkX graph object](https://networkx.org/). Skeletons/Trees are grouped by the provided groups in the NML file.
```
Comment (daniel-wer) [inline docs/wknml.md:311]:
```suggestion
A utility to convert a wK NML object into a [NetworkX graph object](https://networkx.org/). Skeletons/Trees are grouped by the provided groups in the NML file.
```

File: docs/wknml.md [hunk 8]
```diff
@@ -197,3 +229,119 @@ Writes an NML object to a file on disk.
...
+A utility to convert a single wK Tree object into a [NetworkX graph object](https://networkx.org/).
+
+<a name="wknml.nml_generation.extract_nodes_and_edges_from_graph"></a>
+#### extract\_nodes\_and\_edges\_from\_graph
+
+```python
+extract_nodes_and_edges_from_graph(graph: nx.Graph) -> Tuple[List[Node], List[Edge]]
+```
+
+A utility to convert a single [NetworkX graph object](https://networkx.org/) into a list of `Node`objects and `Edge` objects.
```
Comment (daniel-wer) [inline docs/wknml.md:340]:
```suggestion
A utility to convert a single [NetworkX graph object](https://networkx.org/) into a list of `Node` objects and `Edge` objects.
```

File: docs/wknml.md [hunk 9]
```diff
@@ -197,3 +229,119 @@ Writes an NML object to a file on disk.
...
+
+Return
+    Tuple[List[Node], List[Edge]]: A tuple contain both all nodes and edges
```
Comment (daniel-wer) [inline docs/wknml.md:343]:
```suggestion
    Tuple[List[Node], List[Edge]]: A tuple containing all nodes and all edges
```

File: tests/test_nml_objects.py [hunk 10]
```diff
@@ -1,5 +1,7 @@
 from itertools import count
 
+from networkx.algorithms.cuts import volume
+
```
Comment (daniel-wer) [inline tests/test_nml_objects.py:4]:
```suggestion
```

Looks like a superfluous auto-import

**Review by daniel-wer — CHANGES_REQUESTED** [review]
The `<volume>` tag support looks very good to me. Thanks a lot for your work documenting the missing modules! I only noticed too late that I made all my suggestions for the generated docs instead of the source files, so you need to transfer my suggestions manually, sorry for that :disappointed:

File: docs/wknml.md [hunk 11]
```diff
@@ -24,17 +35,22 @@ class NMLParameters(NamedTuple)
 
 Contains common metadata for NML files
 
+**Notes**:
+
+  Setting a task or user bounding boxes will cause wK to 1) render these visually and 2) prevent data loading from outside them.
```
[the remark this answers is not in the document]
Comment (hotzenklotz) [reply docs/wknml.md:40]: Well, it already is super wK-specific is it not? Tasks are very wK. NMLs are very wK.

[the remark this answers is not in the document]
Comment (daniel-wer) [reply docs/wknml.md]: Yes, I see your point. I just wasn't sure whether this information could be better suited for the wK tasks documentation. I don't have a strong opinion on this, your call.

**Comment by daniel-wer** [comment]
@hotzenklotz Just to make sure that you see all the feedback I gave - I had to click "Load More"  to see all comments as there were "hidden conversations".

File: docs/wknml.md [hunk 12]
```diff
@@ -24,17 +35,22 @@ class NMLParameters(NamedTuple)
...
-- `scale` - Vector3
-- `offset` - Optional[Vector3]
-- `time` - Optional[int]
-- `editPosition` - Optional[Vector3]
-- `editRotation` - Optional[Vector3]
-- `zoomLevel` - Optional[float]
-- `taskBoundingBox` - Optional[IntVector6]
-- `userBoundingBox` - Optional[IntVector6]
+- `name` _str_ - Name of a dataset that the annotation is based on. Will cause wK to open the given skeleton annotation with the referenced dataset.
+- `scale` _Vector3_ - Voxel scale of the referenced dataset in nanometers.
```
[the remark this answers is not in the document]
Comment (hotzenklotz) [reply docs/wknml.md:46]: Yes, it would be great to automatically document/export the custom type definition as well.

**Pull request merged**